Introduce Diagnostic Decorator and Testing Utilities

Matejkob opened the pull request:
This PR is part of the efforts related to #2214.

#### Commit 1: Add `DiagnosticDecorator` Protocol, `ANSIDiagnosticDecorator` and `BacisDiagnosticDecorator` Implementations
- Introduce a new protocol, `DiagnosticDecorator`, to standardize the decoration of diagnostic output.
- Implement `ANSIDiagnosticDecorator`, which provides ANSI colorization and severity-based prefixes for enhanced visibility.

#### Commit 2: Introduce Diagnostic Testing Utilities
- Add `DiagnosticTestingUtils.swift` to facilitate more effective unit testing of diagnostics.

ahoppen commented on Sources/SwiftDiagnostics/DiagnosticDecorators/ANSIDiagnosticDecorator.swift:
```diff
@@ -0,0 +1,248 @@
+  ///
+  /// - Returns: A tuple containing:
+  ///   - `highlightedSourceCode`: A string that represents the decorated version of the original source code snippet.
+  ///   - `additionalHighlightedLine`: always with nil.
```
Why do we need to return `additionalHighlightedLine` if it’s always `nil`?

Matejkob replied:
It is `nil` in this implementation because the ANSI codes are able to underline lines directly. The assumption is that in future implementations it will contain characters like `~~~` or `---`.

ahoppen replied:
I would have a slight preference to only add `additionalHighlightedLine` once we actually add a diagnostic annotator that uses it, but I’m fine with keeping it as well.

Matejkob replied:
In the [Enhance diagnostic formatter to include trivia between highlight nodes](#2242) thread, you mentioned the idea of changing `assertSyntaxHighlightRanges` to actually underline the highlighted ranges, either via `DiagnosticDecorator` or a custom function. I think that's a really neat idea. Given that we plan to use `DiagnosticDecorator` in the next PR, I'd opt for keeping it as is for now.

ahoppen commented on Sources/SwiftDiagnostics/DiagnosticDecorators/DiagnosticDecorator.swift:
```diff
@@ -0,0 +1,71 @@
+//
+//===----------------------------------------------------------------------===//
+
+/// Protocol that defines a standard interface for decorating diagnostic output in source code.
+/// This protocol is intended to be used by entities such as ``DiagnosticsFormatter`` and ``GroupedDiagnostics``
```
The abstract and discussion of a docc comment should be separated by a newline

```suggestion
/// Protocol that defines a standard interface for decorating diagnostic output in source code.
///
/// This protocol is intended to be used by entities such as ``DiagnosticsFormatter`` and ``GroupedDiagnostics``
```

ahoppen commented on Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift:
```diff
@@ -0,0 +1,221 @@
+
+/// A typealias representing a location marker.
+///
+/// This string serves to pinpoint the exact location of a particular token in the Abstract Syntax Tree (AST).
```
Some terminology: I try to use AST for the C++ AST tree that’s used in the compiler, which isn’t source-accurate (that’s the abstract part to me). swift-syntax works on SwiftSyntax trees or syntax trees.

ahoppen commented on Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift:
````diff
@@ -0,0 +1,221 @@
+///   return 1
+/// }
+/// ```
+typealias LocationMarker = String
````
What do you think about the following? That way you can really specify which node you want to refer to at a specific position and you could also use `NodeLocator` to specify the highlighted nodes.

```swift
/// Refers to the first node in a syntax tree after `marker` that satisfies `match`.
///
/// For example, if you have
/// ```swift
/// var x: Int
/// 1️⃣func foo() {}
/// ```
///
/// Then `NodeLocator(marker: "1️⃣")` refers to the `CodeBlockItemSyntax` that
/// contains the function declaration and
/// `NodeLocator(marker: "1️⃣", matches: { $0.is(FunctionDeclSyntax.self) })`
/// refers to the function within.
struct NodeLocator {
  let marker: String
  let match: (Syntax) -> Bool = { _ in true }

  func findNode(in tree: Syntax) -> Syntax? {
    // TODO: Implement
  }
}
```

Matejkob replied:
I have a feeling this is going to be a bigger discussion. How about we move it out of the scope of this PR for now and discuss it in the issue: [Introduce Efficient Node Localization in Syntax Tree for Unit Tests](https://github.com/apple/swift-syntax/issues/2251)?

ahoppen replied:
Yes, sure!

ahoppen commented on Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift:
```diff
@@ -0,0 +1,221 @@
+  ///   - fixIts: An array of Fix-It descriptors for quick fixes. Default is an empty array.
+  init(
+    locationMarker: LocationMarker,
+    id: MessageID,
```
How about defaulting this to `.test`? We don’t really care about the ID in these tests.

ahoppen reviewed (commented):
Thanks for splitting the PR. It makes the review a lot easier. And also thanks for the long comments, they also really help understand what’s going on.

I left a couple comments inline. On a high level, I’ve got a couple of questions:
- What do you imagine the use case of implementing a `DiagnosticDecorator` to be? AFAICT `ANSIDiagnosticGenerator` is the only one that we currently have and that you’ll introduce in #2214 and I would prefer to not have the `DiagnosticDecorator` protocol abstraction unless we actually use it.
- I would prefer to not make `DiagnosticDecorator`, `ANSIDiagnosticDecorator` public. If we do, it becomes part of the public API surface and it’s a lot harder to change in the future if we don’t want to break it.
- AFAICT nothing is using the `DiagnosticDecorator` yet. I understand that you will be using it in #2214 but I think it would be good to start using it in the current implementation already unless it’s a huge amount of work. I understand that it’s more work that might seem unnecessary but it would really split #2214 into two self-contained pieces that are easier to review.

Matejkob commented:
Thank you for taking the time to review my PR @ahoppen ; your feedback is invaluable, and I really appreciate the time you've invested.

Addressing your concerns:

> * What do you imagine the use case of implementing a `DiagnosticDecorator` to be? AFAICT `ANSIDiagnosticGenerator` is the only one that we currently have and that you’ll introduce in [Enhancements to diagnostic formatters and testing utilities #2214](https://github.com/apple/swift-syntax/pull/2214) and I would prefer to not have the `DiagnosticDecorator` protocol abstraction unless we actually use it.

The overarching vision for the `DiagnosticDecorator`, as well as other key components like `DiagnosticsFormatter` and `GroupedDiagnostics` in the `SwiftDiagnostics` package, which I have in mind, is to serve as highly customizable and modular tools. The long-term goal would be to empower end-users either by allowing them to build upon these components or by using them directly. To give you a concrete example, @pointfreeco [swift-macro-testing](https://github.com/pointfreeco/swift-macro-testing) library has already adopted the `DiagnosticsFormatter`. Unfortunately, they had to copy-paste the code into their project and manually adjust its implementation, even though their project depends on swift-syntax. This is hardly an ideal scenario and also illustrates the need for a more modular approach.

But it's not just about third-party implementations; we also have potential use-cases within our own ecosystem. For instance, consider the area of macro diagnostics testing. Currently, we have an assert method for macros, which, while functional, lacks advanced diagnostic capabilities. I believe there's a real opportunity here to provide the developer community with a simple yet effective tool for visualizing and debugging macro diagnostics. As we know, debugging macros can be challenging due to the compiler expansions involved, and this feature could significantly alleviate that pain.

Lastly, let's talk about the immediate benefits. We are already planning enhancements to `DiagnosticsFormatter` by adding missing diagnostic elements, which will require robust unit testing. Currently, those unit tests containing diagnostic's highlights are hard to read and maintain, especially when they include complicated diagnostic output expectations. Take this example:

```swift
let expectedOutput = """
  \u{001B}[0;36m1 │\u{001B}[0;0m for \u{001B}[4;39m(i\u{001B}[0;0m \u{001B}[4;39m= 🐮; i != 👩‍👩‍👦‍👦; i += 1)\u{001B}[0;0m { }
    \u{001B}[0;36m│\u{001B}[0;0m │      ╰─ \u{001B}[1;31merror: \u{001B}[1;39mexpected ')' to end tuple pattern\u{001B}[0;0m
    \u{001B}[0;36m│\u{001B}[0;0m ╰─ \u{001B}[1;31merror: \u{001B}[1;39mC-style for statement has been removed in Swift 3\u{001B}[0;0m
  """
```

With the decorator abstraction in place, we can create more readable and developer-friendly unit tests, as shown below:

```swift
let expectedOutput = """
  1 │ for (i = 🐮; i != 👩‍👩‍👦‍👦; i += 1) { }
    │     ~~ ~~~~~~~~~~~~~~~~~~~~~~~
    │ │      ╰─ error: expected ')' to end tuple pattern
    │ ╰─ error: C-style for statement has been removed in Swift 3

  """
```

> * I would prefer to not make `DiagnosticDecorator`, `ANSIDiagnosticDecorator` public. If we do, it becomes part of the public API surface and it’s a lot harder to change in the future if we don’t want to break it.

You bring up a valid concern about making `DiagnosticDecorator` and `ANSIDiagnosticDecorator` public, given that it would lock us into maintaining those APIs and limit our ability to make future changes. I completely agree with you on this. In fact, with the understanding that these APIs are likely to evolve, I've already marked `DiagnosticDecorator` and `ANSIDiagnosticDecorator` with `@_spi(Testing)`. This should help us maintain the flexibility to adapt and refine these components as we continue to build upon them.

> * AFAICT nothing is using the `DiagnosticDecorator` yet. I understand that you will be using it in [Enhancements to diagnostic formatters and testing utilities #2214](https://github.com/apple/swift-syntax/pull/2214) but I think it would be good to start using it in the current implementation already unless it’s a huge amount of work. I understand that it’s more work that might seem unnecessary but it would really split [Enhancements to diagnostic formatters and testing utilities #2214](https://github.com/apple/swift-syntax/pull/2214) into two self-contained pieces that are easier to review.

Great point about immediate use. I've already integrated `DiagnosticDecorator` into `DiagnosticsFormatter` to make this PR more self-contained and easier to review.

> [!note]
> **Disclaimer**:  I want to clarify that my proposed changes stem from research, requirement gathering, and learning about the existing landscape—rather than imposing my own singular vision. I recognize the challenges in fully capturing the current and future needs for diagnostic formatters. The PRs are meant as a collaborative stepping stone, combining what I've learned, observed in other projects, and a few of my own ideas. My aim is for us to build a unified vision together, using the code I've already written as a foundation.

ahoppen commented on Sources/SwiftDiagnostics/DiagnosticDecorators/ANSIDiagnosticDecorator.swift:
```diff
@@ -0,0 +1,248 @@
+/// 2. It has the capability to colorize this prefix and other segments of the message for enhanced visibility
+/// and quick interpretation. This colorization feature can be toggled on or off, offering flexibility in its usage.
+@_spi(Testing) public struct ANSIDiagnosticDecorator: DiagnosticDecorator {
+  /// A flag indicating whether or not to apply colorization.
+  private let colorize: Bool
```
Instead of having a `colorize` flag here, shouldn’t we just have two decorators? One that does add color and one that does not?

Matejkob replied:
Great idea! Thanks! Implemented!

Matejkob replied:
🫶

ahoppen commented on Sources/SwiftDiagnostics/DiagnosticDecorators/DiagnosticDecorator.swift:
```diff
@@ -0,0 +1,72 @@
+/// to `DiagnosticDecorator` to implement custom decorating logic. This allows for different visual representations,
+/// such as using ANSI colors, underscores, emoji-based or other markers, for diagnostics in source code.
+@_spi(Testing) public protocol DiagnosticDecorator {
+  /// Decorates a diagnostic message based on its severity level.
+  ///
+  /// - Parameters:
+  ///   - message: The diagnostic message that needs to be decorated.
+  ///   - severity: The severity level associated with the diagnostic message.
+  ///
+  /// - Returns: A decorated version of the diagnostic message, enhanced by visual cues like color, text styles, or other markers
+  ///            based on its severity level.
```
I think this doc comment should also contain that `decorateMessage` should add any prefix like `error:` or `warning:`. It’s not obvious from the protocol definition.

Matejkob replied:
Is now better? I've added more information to the documentation.

ahoppen replied:
Sounds good 👍🏽

ahoppen commented on Sources/SwiftDiagnostics/DiagnosticDecorators/DiagnosticDecorator.swift:
```diff
@@ -0,0 +1,72 @@
+/// The protocol is designed to be easily customizable. Developers can create their own entities that conform
+/// to `DiagnosticDecorator` to implement custom decorating logic. This allows for different visual representations,
+/// such as using ANSI colors, underscores, emoji-based or other markers, for diagnostics in source code.
+@_spi(Testing) public protocol DiagnosticDecorator {
```
I think we don’t need to make `DiagnosticDecorator` `@_spi(Testing) public` yet, we can just keep it internal (`ANSIDiagnosticDecorator` needs to say `@_spi(Testing)`).

And just to clarify: I think that making `DiagnosticDecorator` might be a very good idea in the future. I would just like to get some experience with the API by implementing a few more diagnostic decorators internally before deciding that we are happy enough with the API to make it public.

Matejkob replied:
Done

Matejkob replied:
I've modified the implementation to enable error throwing. Let me know what you think.

ahoppen commented on Sources/SwiftDiagnostics/GroupedDiagnostics.swift:
```diff
@@ -172,7 +172,7 @@ extension GroupedDiagnostics {
     )
 
-    let colorizeBufferOutline = formatter.colorizeBufferOutline
+    let decorateBufferOutline = formatter.diagnosticDecorator.decorateBufferOutline
```
I would prefer to call `formatter.diagnosticDecorator.decorateBufferOutline` below multiple times instead of storing the function as a closure. If you do want to abbreviate it, I would be fine with storing the diagnostic decorator in a temporary variable. 

Storing functions in variables just makes it a lot harder to understand what’s going on IMO because you can’t do Quick Help on the call and jump to definition requires two jumps instead of one.

```swift
let decorator = formatter.diagnosticDecorator
…
… = decorator.decorateBufferOutline(…)
```

Matejkob replied:
I've added `let diagnosticDecorator = formatter.diagnosticDecorator` instead of storing func inn closure.

ahoppen commented on Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift:
```diff
@@ -0,0 +1,221 @@
+  ///   - file: The file where the test is located, used for reporting failures.
+  ///   - line: The line where the test is located, used for reporting failures.
+  ///
+  /// - Returns: A ``Diagnostic`` instance populated with details from the ``DiagnosticDescriptor``, or `nil` if it fails.
```
Instead of silently failing and not generating a `Diagnostic` for a `DiagnosticDescriptor` if it fails, what do you think about throwing an error instead? I think that should make it a lot clearer what’s going wrong if you write eg. a `DiagnosticDescriptor` with a marker that doesn’t exist in the source code.

Matejkob replied:
I believe that implementing the `assert` function in this way would indeed make the code clearer, but I'm not entirely convinced that the end result would be more transparent to the users of this high-level API. 

My primary motivation behind this implementation approach was to avoid interrupting the assertion process unnecessarily. The `assertAnnotated` function takes a collection of `diagnosticDescriptors` as an argument. I chose to return `nil` while also invoking `XCTFail` so that we could iterate over all elements, inform the user only about the missing ones, and apply the existing ones to provide as much assertion validation as possible. For instance, if only one location marker out of four is missing, the user will end up with the following result:

![Screenshot 2023-10-11 at 7 18 50 PM](https://github.com/apple/swift-syntax/assets/44930823/d1db24f0-6def-4155-9dd8-1a054c67f4e7)

I believe this approach benefits end clients. If I'm not mistaken, it wouldn't work in the same way if we chose to throw an error, would it?

ahoppen replied:
Oh, I missed the fact that we always emitted an `XCTFail` if we return `nil`.

And I think me missing that is a reason why I would prefer to `throw`. My thinking is that if you are missing one of the diagnostic markers in your test input, there’s something wrong with the test case itself (not just a runtime failure) and thus early exiting is quite alright.

As it stands right now, every time I would read this code, I would wonder the same thing that I originally commented: 
- Are we silently dropping one of our test case assertions here? That seems dangerous
- Looks at implementation/documentation: Oh, no, it’s fine we emit an `XCTFail` in that case.
- Now, what was I doing originally? Needs to restore mental context after side-stepping.

Matejkob replied:
You make some solid points. You got me—I'm convinced.

I'll swap out the `nil` and `XCTFail` combo for a good ol' error throw. Thanks for the nudge!

ahoppen commented on Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift:
```diff
@@ -0,0 +1,221 @@
+
+    guard let diagnosticNode = node(at: self.locationMarker) else { return nil }
+
+    var notes = [Note]()
+    for noteDescriptor in self.noteDescriptors {
+      guard let noteNode = node(at: noteDescriptor.locationMarker) else { continue }
+
+      let note = Note(
+        node: noteNode,
+        message: SimpleNoteMessage(message: noteDescriptor.message, fixItID: noteDescriptor.id)
+      )
+      notes.append(note)
+    }
```
I think using `map` here might make it a little bit more swifty + easier to understand, especially if `node(at:)` throws on failure.

Matejkob replied:
My idea was to avoid stoping the assertion process if just one of the nodes couldn't be found. That's why I've used a `for` loop instead of `map`, as I'm employing `continue` to skip to the next iteration. However, I believe we could also use `compactMap` as an alternative.

Matejkob replied:
Changed to `map` – done.

ahoppen commented on Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift:
```diff
@@ -0,0 +1,221 @@
+  let (markers, source) = extractMarkers(markedSource)
+  let tree = Parser.parse(source: source)
+
+  // Check for any unexpected diagnostics generated by parser.
+  let generatedDiagnostics = ParseDiagnosticsGenerator.diagnostics(for: tree)
+  if !generatedDiagnostics.isEmpty {
+    let diagnosticMessages = generatedDiagnostics.map { $0.message }.joined(separator: ", ")
+    XCTFail("Unexpected diagnostics found in the source: \(diagnosticMessages)", file: file, line: line)
+  }
```
Why do we need to check this? I think it would be totally fine to add our own diagnostics based on the `DiagnosticDescriptor`s even if the syntax tree contains errors.

Matejkob replied:
Sounds good to me!

Matejkob replied:
Removed

ahoppen commented on Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift:
```diff
@@ -35,34 +25,22 @@ extension GroupedDiagnostics {
-      )
-      diagnostics.append(diag)
+
+    let parserDiagnostics = ParseDiagnosticsGenerator.diagnostics(for: tree)
```
Regarding my comment above about why we need to check that the tree didn’t contain any errors, I think if we removed this line as well and only displayed errors based on `diagnosticDescriptors`, we would be completely independent of `SwiftParserDiagnostics` here.

Matejkob replied:
I love the idea, but I'd prefer to hold off on that for this PR. `GroupedDiagnostics` has its own set of issues, and if we were to remove this part now, we'd likely find ourselves fixing additional problems. After I've improved `DiagnosticsFormatter`, I plan on revisiting `GroupedDiagnostics`. I'll definitely tackle it at that point.

ahoppen replied:
OK sounds good to me.

ahoppen reviewed (commented):
Thanks for providing the detailed response. I think the overall direction is very good. I would just like to keep `DiagnosticDecorator` internal until we have gained some more experience with it. I think all the other comments are fairly minor.

ahoppen commented on Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:
```diff
@@ -73,12 +73,15 @@ public struct DiagnosticsFormatter {
   /// Number of lines which should be printed before and after the diagnostic message
   public let contextSize: Int
 
-  /// Whether to colorize formatted diagnostics.
-  public let colorize: Bool
```
The removal of the `colorize` property is an API-breaking change. 

To maintain source compatibility, could you add

```swift
@available(*, deprecated, message: "Store the `colorize` property passed to the initializer instead")
public var  colorize: Bool {
  return diagnosticDecorator is ANSIDiagnosticDecorator
}

Matejkob replied:
Done

ahoppen commented on Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift:
```diff
@@ -0,0 +1,224 @@
+    let diagnosticNode = try node(at: self.locationMarker)
+
+    let notes = try self.noteDescriptors.map { noteDescriptor in
+      let noteNode = try node(at: noteDescriptor.locationMarker)
+
+      return Note(
+        node: noteNode,
+        message: SimpleNoteMessage(message: noteDescriptor.message, fixItID: noteDescriptor.id)
+      )
```
Nit: `noteNode` is a weird name and we don’t actually need it.

```suggestion
      return Note(
        node: try node(at: noteDescriptor.locationMarker),
        message: SimpleNoteMessage(message: noteDescriptor.message, fixItID: noteDescriptor.id)
      )
```

Matejkob replied:
Great idea! Done

ahoppen reviewed (commented):
One really small nitpick, otherwise looks good to me 👍🏽

ahoppen commented:
@swift-ci Please test

Matejkob commented:
CI failed with the following message:

> DiagnosticsFormatter.swift:80:28: error: cannot find type 'DiagnosticDecorator' in scope

From a Swift perspective, this doesn't seem to be failing, since it's in the same module and the type is internal. Should I modify `CMakeLists.txt`? Is there any documentation where I can learn more about what's going on here?

bnbarham commented:
> Should I modify CMakeLists.txt?

Yeah, you'll need to add all the new files to cmake as well.

> Is there any documentation where I can learn more about what's going on here?

Not really. The compiler uses cmake rather than SPM to build swift-syntax (for the libraries included in the toolchain), so any new files have to be added to the relevant CMakeLists.txt.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows

ahoppen commented:
@swift-ci Please test Windows

ahoppen commented:
```
Test Case '-[SwiftDiagnosticsTest.GroupedDiagnosticsFormatterTests testGroupingForMacroExpansion]' failed (0.487 seconds).
```
@Matejkob Could you look at the failure?

Matejkob commented:
Oops! I mistakenly committed a modified test and disabled the `GroupedDiagnosticsFormatterTests` test case in Xcode. Consequently, when I ran all the tests, it didn't fail. My apologies for the oversight. 🙈

ahoppen commented:
No problem. Let’s try again.

@swift-ci Please test

Matejkob commented:
Is this the main problem `/Users/ec2-user/jenkins/workspace/swift-syntax-PR-macOS/branch-main/swift/lib/LLVMPasses/ARCEntryPointBuilder.h:21:10: fatal error: 'llvm/TargetParser/Triple.h' file not found
`?

ahoppen commented:
Seems like an unrelated issue. Trying again.

@swift-ci Please test macOS

ahoppen commented:
@swift-ci Please test Windows
